src/main/java/org/tugraz/sysds/runtime/instructions/fed/InitFEDInstruction.java

```diff
@@ -34,106 +35,126 @@
-					for (int d = 0; d < beginDims.length; d++) {
-						beginDims[d] = ((ScalarObject) beginDimsData.get(d)).getLongValue();
-						endDims[d] = ((ScalarObject) endDimsData.get(d)).getLongValue();
-					}
-					usedDims[0] = Math.max(usedDims[0], endDims[0]);
-					usedDims[1] = Math.max(usedDims[1], endDims[1]);
-					try {
-						FederatedData federatedData = new FederatedData(
-								new InetSocketAddress(InetAddress.getByName(host), port), filepath);
-						feds.add(new ImmutablePair<>(new FederatedRange(beginDims, endDims), federatedData));
-					}
-					catch (UnknownHostException e) {
-						throw new DMLRuntimeException("federated host was unknown: " + host);
-					}
+			if (addressData instanceof StringObject) {
+
+				// We split address into url/ip, the port and file path of file to read
+				String[] parsedValues = parseURL(((StringObject) addressData).getStringValue());
+				String host = parsedValues[0];
+				int port = Integer.parseInt(parsedValues[1]);
+				String filePath = parsedValues[2];
+				// get beginning and end of data ranges
+				List<Data> rangesData = ranges.getData();
+				Data beginData = rangesData.get(i * 2);
+				Data endData = rangesData.get(i * 2 + 1);
+				if (beginData.getDataType() != Types.DataType.LIST || endData.getDataType() != Types.DataType.LIST)
+					throw new DMLRuntimeException(
+							"Federated read ranges (lower, upper) have to be lists of dimensions");
+				List<Data> beginDimsData = ((ListObject) beginData).getData();
+				List<Data> endDimsData = ((ListObject) endData).getData();
+
+				// fill begin and end dims
+				long[] beginDims = new long[beginDimsData.size()];
+				long[] endDims = new long[beginDims.length];
+				for (int d = 0; d < beginDims.length; d++) {
+					beginDims[d] = ((ScalarObject) beginDimsData.get(d)).getLongValue();
+					endDims[d] = ((ScalarObject) endDimsData.get(d)).getLongValue();
 				}
-				else {
-					throw new DMLRuntimeException("federated address `" + address + "` does not fit required pattern " +
-							"of \"host:port/directory\"");
+				usedDims[0] = Math.max(usedDims[0], endDims[0]);
+				usedDims[1] = Math.max(usedDims[1], endDims[1]);
+				try {
+					FederatedData federatedData = new FederatedData(
+							new InetSocketAddress(InetAddress.getByName(host), port), filePath);
+					feds.add(new ImmutablePair<>(new FederatedRange(beginDims, endDims), federatedData));
+				} catch (UnknownHostException e) {
+					throw new DMLRuntimeException("federated host was unknown: " + host);
 				}
-			}
-			else {
+
+			} else {
 				throw new DMLRuntimeException("federated instruction only takes strings as addresses");
 			}
 		}
 		MatrixObject output = ec.getMatrixObject(_output);
 		output.getDataCharacteristics().setRows(usedDims[0]).setCols(usedDims[1]);
 		federate(output, feds);
 	}
-	
+
+	public static String[] parseURL(String input) {
+		try {
+			// Artificially making it http protocol. It is not used in the end.
```

Comment:
Add the reason for the workaround (e.g., that URL does not accept input without protocol specifier and you didn't want to define a new proto at this time)

Reply:
done